pennylane/io/qualtran_io.py

```diff
@@ -0,0 +1,428 @@
+# Copyright 2018-2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule contains the adapter class for Qualtran-PennyLane interoperability.
+"""
+from collections import defaultdict
+from functools import lru_cache, singledispatch
+
+import numpy as np
+
+import pennylane as qml
+from pennylane.operation import DecompositionUndefinedError, MatrixUndefinedError, Operation
+from pennylane.wires import WiresLike
+
+try:
+    import qualtran as qt
+except (ModuleNotFoundError, ImportError) as import_error:
+    pass
+
+
+# TODO: Delete this when new version of Qualtran is released
+# pylint: disable=unused-argument
+@lru_cache
+def _get_to_pl_op():
+
+    @singledispatch
+    def _to_pl_op(bloq, wires):
+        return FromBloq(bloq=bloq, wires=wires)
+
+    @_to_pl_op.register
+    def _(bloq: qt.bloqs.basic_gates.CNOT, wires):
+        return qml.CNOT(wires=wires)
+
+    @_to_pl_op.register
+    def _(bloq: qt.bloqs.basic_gates.GlobalPhase, wires):
+        return qml.GlobalPhase(bloq.exponent * np.pi, wires)
+
+    @_to_pl_op.register
+    def _(bloq: qt.bloqs.basic_gates.Hadamard, wires):
+        return qml.Hadamard(wires)
+
+    @_to_pl_op.register
+    def _(bloq: qt.bloqs.basic_gates.Identity, wires):
+        return qml.Identity(wires)
+
+    @_to_pl_op.register
+    def _(bloq: qt.bloqs.basic_gates.Rx, wires):
+        return qml.RX(bloq.angle, wires)
+
+    @_to_pl_op.register
+    def _(bloq: qt.bloqs.basic_gates.Ry, wires):
+        return qml.RY(bloq.angle, wires)
+
+    @_to_pl_op.register
+    def _(bloq: qt.bloqs.basic_gates.Rz, wires):
+        return qml.RZ(bloq.angle, wires)
```

Comment:
These don't require `\pi` scaling like `GlobalPhase`? Also how are `PowGates` being handled?

Reply:
No, the implementation for RX RY RZ follow the Cirq implementation, which is equivalent to PL's, so there's no pi scaling. PowGates are handled by the default FromBloq wrapping, we don't have a specific equivalent in PL.